libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,192 @@
+			output += collect(declaration, _sourceUnit, identifier->name());
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		output += collect(identifierPath->annotation().referencedDeclaration, _sourceUnit, identifierPath->path().back());
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_sourceNode))
+		output += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+		output += collect(declaration, _sourceUnit, declaration->name());
+	else
+		lspAssert(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
+
+	return output;
+}
+
+bool ReferenceCollector::visit(ImportDirective const& _import)
+{
+	if (_import.name() == m_sourceIdentifierName)
+		return true;
+	return false;
+}
+
+void ReferenceCollector::endVisit(ImportDirective const& _import)
+{
+	for (auto const& symbolAlias: _import.symbolAliases())
+	{
+		if (
+			m_sourceIdentifierName == *symbolAlias.alias &&
+			symbolAlias.symbol &&
+			symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration
+		)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Read);
+			break;
+		}
+		else if (m_sourceIdentifierName == *symbolAlias.alias)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Text);
+			break;
+		}
+	}
+}
+
+bool ReferenceCollector::tryAddReference(Declaration const* _declaration, SourceLocation const& _location)
+{
+	if (&m_declaration != _declaration)
+		return false;
+
+	m_resultingReferences.emplace_back(_location, m_kind);
+	return true;
+}
+
+void ReferenceCollector::endVisit(Identifier const& _identifier)
+{
+	if (auto const* declaration = _identifier.annotation().referencedDeclaration)
+		tryAddReference(declaration, _identifier.location());
+
+	for (auto const* declaration: _identifier.annotation().candidateDeclarations + _identifier.annotation().overloadedDeclarations)
+		tryAddReference(declaration, _identifier.location());
+}
+
+
+void ReferenceCollector::endVisit(IdentifierPath const& _identifierPath)
+{
+	tryAddReference(_identifierPath.annotation().referencedDeclaration, _identifierPath.location());
+}
+
+void ReferenceCollector::endVisit(MemberAccess const& _memberAccess)
+{
+	if (_memberAccess.annotation().referencedDeclaration == &m_declaration)
+		m_resultingReferences.emplace_back(_memberAccess.location(), m_kind);
+}
+
+bool ReferenceCollector::visit(Assignment const& _node)
+{
+	auto const restoreKind = ScopeGuard{[this, savedKind=m_kind]() { m_kind = savedKind; }};
+
+	m_kind = DocumentHighlightKind::Write;
+	_node.leftHandSide().accept(*this);
+
+	m_kind = DocumentHighlightKind::Read;
+	_node.rightHandSide().accept(*this);
+
+	return false;
+}
```

Comment:
it won't. and we decided to not do it vis-a-vis :)

Reply:
Didn't we actually decide to use the AST annotations for deciding whether it's a write or read and drop the simplistic "write for left side, read for right side"? That's what i remember from that discussion.

We have the [`willBeWrittenTo`](https://github.com/ethereum/solidity/blob/v0.8.17/libsolidity/ast/ASTAnnotations.h#L274-L275) annotation in the AST. This might be the one Daniel meant.

Reply:
And whatever the decision, would be best to have a test case documenting it.

Reply:
@cameel IIRC Daniel only talked about the possibility. But it's done now. :)
Read/Write is tested already, what more do you need a test case for assignments?

Reply:
I meant the test case for `x.push() = 1`. Whether we handle it correctly or not, we should have a case that shows what the current expected behavior is.